Functor smoothing

GiudGiud opened the pull request:
before and after on the MSFR initial no-thermal velocity, using average interpolation for velocity, and smoothing the output pressure field. We could use the output pressure field for evaluating matprops too?

<img width="655" alt="Screenshot 2023-09-13 at 1 43 57 PM" src="https://github.com/idaholab/moose/assets/25037151/2028ae84-8146-420b-8f12-abc456e1bec7">
<img width="655" alt="Screenshot 2023-09-13 at 1 50 17 PM" src="https://github.com/idaholab/moose/assets/25037151/564432e2-9903-498b-b345-a190a8ac15f9">

so p_elem = p_elem / 2 + p_max_neighbor / 4 + p_opposite_of_max_neighbor / 4

@grmnptr  is this a crime against CFD

GiudGiud committed:
Add functor smoother to framework with 3 heuristics
refs #25452

GiudGiud committed:
Add test for functor smoother

GiudGiud committed:
Add support for parallelism for functor smoothing

GiudGiud committed:
Add documentation for functor smoother

GiudGiud commented on framework/doc/content/source/functormaterials/FunctorSmoother.md:
```diff
@@ -0,0 +1,35 @@
+- neighbor averaging
+
+The elemental value of the smoothed functor is equal to the arithmetic average of the values of the previous
+functor at the neighbors of the element.
```
This one I could remove and instead rely on the smoothing layers treatment

GiudGiud replied:
I ended up doing my thing just because it was the same code structure as the other. Figuring out how to make the general smoothing layers treatment would have taken me a little more time

GiudGiud commented:
I ll get the pressure drop see what impact this does

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1766662/) on aec313a wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/25455/site)

This comment will be updated on new commits.

grmnptr commented:
I like the idea of a functor smoother but my first opinion is that using if to handle the checker boarding is a crime against finite volume CFD. Let's see what the comparison between RC and the smoothed version says.

GiudGiud commented:
--------------------- ----------  mdot (core middle)       Pdrop (around HX + pump)
Average + smoothing P                 2.331049e+03.  -7.101159e+04
RC no smoothing.                         8.901196e+03   -6.899656e+04
Average + smoothing P and vx/y  2.331049e+03 -7.101159e+04

so while it does good on pressure drop, the velocities are quite off
It's also unfortunate to have to smooth them as well. The smoothing does a good job, but the argument is a lot weaker for smoothing them (they do play a role, and smoothing them should lead to a field that conserves mass and momentum otherwise why did we bother)

so until we get better results, this will just have to be a functor smoother not a trick in NS

GiudGiud commented:
do we expect velocity to checkerboard with average interpolation and not just pressure (as I remember only pressure) ?

if not, does anyone have another case that exhibits checkerboarding?

lindsayad commented:
> do we expect velocity to checkerboard with average interpolation

It depends what interpolation you're talking about. `advected_interp_method = average` can lead to oscillations when advection is significant. But I do not expect oscillations in the velocity field just from `velocity_interp_method = average`

lindsayad commented:
Your table concerns me in the opposite way. I am worried that the Rhie-Chow velocity field is more wrong than the average velocity field. I'm curious what your figure of merit converges to with mesh refinement

lindsayad commented:
Applied to `advected_interp_method`, `velocity_interp_method`, both?

GiudGiud commented:
advected_interp_method is upwind (default)

lindsayad commented:
Hmm interesting. I may have heard in the past that you can see checkerboarding in the velocity field with an average advecting velocity interpolation, but I forget why that can happen

grmnptr commented:
We have two kinds of checkerboarding:
1. An oscillation coming from the non-boundedness (or boundlessness?) of the advected interpolation scheme. Average is not bounded so in many cases we see this. Upwind is bounded so it would not show this oscillation.
2. The interaction between linear interpolation in the Green-Gauss gradient of the pressure gradient and a linear velocity interpolation method in the continuity equation. This is what RC is for. I think the RC results are correct and the smoothed version are not. In multiD the interaction can be pretty complex so I am not sure if the smoothing of the fields will give you the right guess for the solution.

GiudGiud commented:
Results
<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:////Users/giudgl/Library/Group%20Containers/UBF8T346G9.Office/TemporaryItems/msohtmlclip/clip.htm">
<link rel=File-List
href="file:////Users/giudgl/Library/Group%20Containers/UBF8T346G9.Office/TemporaryItems/msohtmlclip/clip_filelist.xml">
<style>
<!--table
	{mso-displayed-decimal-separator:"\.";
	mso-displayed-thousand-separator:"\,";}
@page
	{margin:.75in .7in .75in .7in;
	mso-header-margin:.3in;
	mso-footer-margin:.3in;}
tr
	{mso-height-source:auto;}
col
	{mso-width-source:auto;}
br
	{mso-data-placement:same-cell;}
td
	{padding-top:1px;
	padding-right:1px;
	padding-left:1px;
	mso-ignore:padding;
	color:black;
	font-size:12.0pt;
	font-weight:400;
	font-style:normal;
	text-decoration:none;
	font-family:Calibri, sans-serif;
	mso-font-charset:0;
	mso-number-format:General;
	text-align:general;
	vertical-align:bottom;
	border:none;
	mso-background-source:auto;
	mso-pattern:auto;
	mso-protection:locked visible;
	white-space:nowrap;
	mso-rotate:0;}
.xl63
	{mso-number-format:Scientific;}
-->
</style>
</head>

<body link="#0563C1" vlink="#954F72">



  | RC |   | average +   pressure smoothing
-- | -- | -- | --
  | mdot_hx | pdrop | mdot_hx | pdrop |  
r0 | 1.70E+04 | 6.88E+04 | 1.02E+04 | 7.10E+04 |  
r1 | 1.89E+04 | 6.77E+04 | 1.10E+04 | 7.01E+04 |  
r2 | no CV |   | 1.33E+04 | 6.95E+04 |  



</body>

</html>

GiudGiud commented:
How are we feeling about merging this?
I feel it s not a bad idea to be able to smooth functions or variables, and the checkerboard removing is, while useless right now, quite impressive?

lindsayad commented on framework/src/functormaterials/FunctorSmoother.C:
```diff
@@ -0,0 +1,219 @@
+
+              // Find the neighbor with the furthest value
+              typename Moose::ADType<T>::type extreme_value = 0;
+              typename Moose::ADType<T>::type distance = 0;
```
I think I'd name this something like `delta`

GiudGiud replied:
done

lindsayad commented on framework/src/functormaterials/FunctorSmoother.C:
```diff
@@ -0,0 +1,219 @@
+                }
+              }
+
+              if (r_elem->n_neighbors() == 1)
```
there are no elements in libMesh with this number of neighbors

lindsayad replied:
Note that this API won't check for null neighbors

GiudGiud replied:
I dont like that much. Seems like a trap

lindsayad replied:
Take it up with @roystgnr 😄

roystgnr replied:
`class Circle : public Face {` ...

lindsayad commented on framework/src/functormaterials/FunctorSmoother.C:
```diff
@@ -0,0 +1,219 @@
+
+              if (r_elem->n_neighbors() == 1)
+              {
+                average += extreme_value / 2;
+                return average;
+              }
+              else
+                average += extreme_value / 4;
```
I don't understand the choice of divisor. I'd appreciate a comment

GiudGiud replied:
added some text

lindsayad commented on framework/src/functormaterials/FunctorSmoother.C:
```diff
@@ -0,0 +1,219 @@
+          {
+            if constexpr (std::is_same_v<const Moose::ElemArg &, decltype(r)>)
+            {
+              unsigned int n_faces = 0;
+              for (const auto side_index : r_elem->side_index_range())
+              {
+                Moose::FaceArg face_arg{_mesh.faceInfo(r_elem, side_index),
+                                        Moose::FV::LimiterType::CentralDifference,
+                                        true,
+                                        false,
+                                        nullptr};
+                if (face_arg.fi)
+                {
+                  average += functor_in(face_arg, t);
+                  n_faces++;
+                }
+              }
+              average /= n_faces;
```
I believe that this is exactly what is being done conceptually in `Reconstructions.h` for `interpolateReconstruct` which is what is used for the porosity smoothing. Perhaps this logic could be factored out although I wouldn't require it

GiudGiud replied:
yes. Am I correct that interpolateReconstruct is meant to build the reconstructed functor once and then provide it as a map functor? whereas here it s more on-the-fly (unless you use caching)

lindsayad replied:
yes that's right

GiudGiud commented on framework/src/functormaterials/FunctorSmoother.C:
```diff
@@ -0,0 +1,219 @@
+        Moose::RelationshipManagerType::COUPLING;
+    rm_params.template set<unsigned short>("layers") = 1;
+    rm_params.template set<bool>("use_point_neighbors") = false;
+    rm_params.template set<bool>("attach_geometric_early") = false;
```
I think it will work as long as some other object asks for the ghosting

lindsayad replied:
It will work as long as some other object has requested late geometric ghosting. Otherwise it will not work

GiudGiud replied:
ah looks like it does not

GiudGiud committed:
Address Alex's review:
- dont add early ghosting since it s too late at algebraic object construction
- distance -> delta
- add comment on choice
- dont trust n_neighbor api that counts null neigbors

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1786488/) on 3779ea8 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>81a54c</th>
      <th colspan=4>#25455 3779ea</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.68%</td>
      <td>85.67%</td>
      <td>-0.01%</td>
      <td>81.82%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>91590</td>
      <td>91658</td>
      <td>+68</td>
      <td>72</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15308</td>
      <td>15328</td>
      <td>+20</td>
      <td>16</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/25455/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/25455/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/25455/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/25455/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/25455/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/25455/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/25455/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/25455/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/25455/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/25455/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/25455/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/25455/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/25455/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/25455/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/25455/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/25455/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/25455/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/25455/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/25455/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/25455/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/25455/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/25455/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/25455/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/25455/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/25455/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/25455/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/25455/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/25455/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/25455/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/25455/coverage/xfem)

</details>

## Warnings

- `framework` new line coverage rate 81.82% is less than the suggested 90.0%



This comment will be updated on new commits.

lindsayad commented:
The OpenMPI failure needs to be addressed

GiudGiud committed:
Add support for distributed meshes by trying both elem and neighbor for the FI

lindsayad commented on framework/src/functormaterials/FunctorSmoother.C:
```diff
@@ -0,0 +1,225 @@
+    rm_params.template set<Moose::RelationshipManagerType>("rm_type") =
+        Moose::RelationshipManagerType::GEOMETRIC | Moose::RelationshipManagerType::ALGEBRAIC |
+        Moose::RelationshipManagerType::COUPLING;
+    rm_params.template set<unsigned short>("layers") = 2 + (_smoothing_technique == FACE_AVERAGE);
```
I don't understand why there needs to be more ghosting for the face averaging technique

GiudGiud replied:
oh this is residue from testing

GiudGiud replied:
looks like we pass testing, lemme remove that

GiudGiud commented:
Thanks for the review!
